src/targeting/personalised.ts

```diff
@@ -0,0 +1,256 @@
+const adManagerGroups = [
+	'1',
+	'2',
+	'3',
+	'4',
+	'5',
+	'6',
+	'7',
+	'8',
+	'9',
+	'10',
+	'11',
+	'12',
+] as const;
+type AdManagerGroup = typeof adManagerGroups[number];
+
+/**
+ * Personalised Targeting requires user consent
+ *
+ * It allows or prevents personalised advertising, restrict data processing
+ * and handles access to cookies and localstorage
+ */
+export type PersonalisedTargeting = {
+	/**
+	 * **A**d **M**anager **T**argeting **Gr**ou**p** – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * Sample values:
+	 * -
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=12318099
+	 * */
+	amtgrp: AdManagerGroup | null;
+
+	/**
+	 * Interaction with TCFv2 banner – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=12083384
+	 */
+	cmp_interaction?: TCEventStatusCode | NotApplicable;
+
+	/**
+	 * **TCFv2 Consent** to [all purposes] – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * [all purposes]: https://vendor-list.consensu.org/v2/vendor-list.json
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=12080297
+	 * */
+	consent_tcfv2: True | False | NotApplicable;
+
+	/**
+	 * **Fr**equency – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=214647
+	 */
+	fr: typeof frequency[number];
+
+	/**
+	 * **P**ersonalised **A**ds Consent – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=11701767
+	 */
+	pa: True | False;
+
+	/**
+	 * **Permutive** user segments – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * Values: 900+ number IDs
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=11958727
+	 */
+	permutive: string[];
+
+	/**
+	 * **R**estrict **D**ata **P**rocessing Flag – [see on Ad Manager][gam]
+	 *
+	 * Type: _Predefined_
+	 *
+	 * [gam]: https://admanager.google.com/59666047#inventory/custom_targeting/detail/custom_key_id=11701767
+	 */
+	rdp: True | False | NotApplicable;
+};
+
+/* -- Methods -- */
+
+const getRawWithConsent = (key: string, state: ConsentState): string | null => {
+	if (state.tcfv2) {
+		if (state.tcfv2.consents['1']) return storage.local.getRaw(key);
+	}
+	if (state.ccpa) {
+		if (!state.ccpa.doNotSell) return storage.local.getRaw(key);
+	}
+	if (state.aus) {
+		if (state.aus.personalisedAdvertising) return storage.local.getRaw(key);
+	}
+
+	return null;
+};
+
+const getFrequencyValue = (
+	state: ConsentState,
+): PersonalisedTargeting['fr'] => {
+	const rawValue = getRawWithConsent('gu.alreadyVisited', state);
+	if (!rawValue) return '0'; // TODO: should we return `null` instead?
```

Comment:
```suggestion
	if (!rawValue) return '0';
```
I think personally this is fine as we aren't exporting this function and it's only being used directly in `getPersonalisedTargeting`. And we can always refactor it if necessary later without needing the `TODO: // ...`